torchtune/modules/attention.py

```diff
@@ -205,65 +209,78 @@ def forward(
         TODO:
             - Return the attention weights
         """
-        # input has shape [b, s_x, d]
+        # x has shape [b, s_x, d]
+        # y has shape [b, s_y, d]
         b, s_x, _ = x.shape
-        y = y if y is not None else x
-        s_y = y.shape[1]
+        s_y = y.shape[1] if y is not None else 0
 
         if self.kv_cache and input_pos is None:
             cache_size = self.kv_cache.size
             input_pos = torch.arange(cache_size, cache_size + s_y, device=x.device)
 
         # q has shape [b, s_x, num_heads * head_dim]
-        # k has shape [b, s_y, num_kv_heads * head_dim]
-        # v has shape [b, s_y, num_kv_heads * head_dim]
         q = self.q_proj(x)
-        k = self.k_proj(y)
-        v = self.v_proj(y)
 
         # number of queries per key/value
         q_per_kv = self.num_heads // self.num_kv_heads
-
-        # q: [b, s_x, n_kv, q_per_kv, h_d]
-        # k: [b, s_y, n_kv, 1, h_d]
-        # v: [b, s_y, n_kv, 1, h_d]
-        q = q.view(b, s_x, self.num_kv_heads, q_per_kv, self.head_dim)
-        k = k.view(b, s_y, self.num_kv_heads, 1, self.head_dim)
-        v = v.view(b, s_y, self.num_kv_heads, 1, self.head_dim)
-
-        # if needed, expand the key and value tensors to have the same shape
-        # as the query tensor by copying values across the relevant dim
-        if self.num_heads != self.num_kv_heads:
-            k = k.expand(b, s_y, self.num_kv_heads, q_per_kv, self.head_dim)
-            v = v.expand(b, s_y, self.num_kv_heads, q_per_kv, self.head_dim)
-
-        # llama applies the RoPE embeddings on tensors with shape
-        # [b, s, n_h, h_d]
-        # Reshape the tensors before we apply RoPE
-        q = q.reshape(b, s_x, -1, self.head_dim)
-        k = k.reshape(b, s_y, -1, self.head_dim)
-        v = v.reshape(b, s_y, -1, self.head_dim)
+        q = q.view(b, s_x, self.num_kv_heads * q_per_kv, self.head_dim)
 
         # Apply positional embeddings
         if self.pos_embeddings is not None:
             q = self.pos_embeddings(q, input_pos=input_pos)
-            k = self.pos_embeddings(k, input_pos=input_pos)
 
-        # [b, n_h, s, h_d]
+        # [b, n_h, s_x, h_d]
         q = q.transpose(1, 2)
-        k = k.transpose(1, 2)
-        v = v.transpose(1, 2)
 
-        # Normalize q and k
-        # Note: normalizing before kv_cache assumes token level
-        #       normalization, not batch or sequence level
+        # Normalize q
         if self.q_norm is not None:
             q = self.q_norm(q)
-            k = self.k_norm(k)
 
-        # Update key-value cache
-        if self.kv_cache is not None:
-            k, v = self.kv_cache.update(input_pos, k, v)
+        if y is None:
+            if self.kv_cache is None:
+                raise ValueError(
+                    "Must provide y input or use kv_cache to enable streaming decoding"
+                )
+            k = self.kv_cache.k_cache
+            v = self.kv_cache.v_cache
+        else:
```

Comment:
add a comment to say that we are doing K and V projections and optionally applying RoPE here since K and V are not cached (basically something to offset the hit to readability by splitting Q logic from K and V logic)